sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PathFilterSuite.scala

```diff
@@ -0,0 +1,593 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+
+import java.io.File
+import java.time.{LocalDateTime, ZoneOffset}
+import java.time.format.DateTimeFormatter
+
+import scala.util.Random
+
+import org.apache.spark.sql.{AnalysisException, QueryTest, Row}
+import org.apache.spark.sql.catalyst.util.{stringToFile, DateTimeUtils}
+import org.apache.spark.sql.test.SharedSparkSession
+
+class PathFilterSuite extends QueryTest with SharedSparkSession {
+  import testImplicits._
+
+  def createSingleFile(dir: File): File = {
+    val file = new File(dir, "temp" + Random.nextInt(1000) + ".csv")
+    stringToFile(file, "text")
+  }
+
+  def setFileTime(time: LocalDateTime, file: File): Boolean = {
+    val sameTime = time.toEpochSecond(ZoneOffset.UTC)
+    file.setLastModified(sameTime * 1000)
+  }
+
+  def setPlusFileTime(time: LocalDateTime, file: File, interval: Long): Boolean = {
+    val sameTime = time.plusDays(interval).toEpochSecond(ZoneOffset.UTC)
+    file.setLastModified(sameTime * 1000)
+  }
+
+  def setMinusFileTime(time: LocalDateTime, file: File, interval: Long): Boolean = {
+    val sameTime = time.minusDays(interval).toEpochSecond(ZoneOffset.UTC)
+    file.setLastModified(sameTime * 1000)
+  }
+
+  def formatTime(time: LocalDateTime): String = {
+    time.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss"))
+  }
+
+  test(
```

Comment:
The overall new tests have similar structure, which can be extracted out.

```
  private def executeTest(
      dir: File,
      fileDate: LocalDateTime,
      modifiedBefore: Option[String] = None,
      modifiedAfter: Option[String] = None,
      pathGlobFilter: Option[String] = None): DataFrame = {
    executeTest(dir, Seq(fileDate), modifiedBefore, modifiedAfter, pathGlobFilter)
  }

  private def executeTest(
      dir: File,
      fileDates: Seq[LocalDateTime],
      modifiedBefore: Option[String] = None,
      modifiedAfter: Option[String] = None,
      pathGlobFilter: Option[String] = None): DataFrame = {
    fileDates.foreach { fileDate =>
      val file = createSingleFile(dir)
      setFileTime(fileDate, file)
    }

    var dfReader = spark.read.format("csv").option("timeZone", "UTC")
    modifiedBefore.foreach { opt => dfReader = dfReader.option("modifiedBefore", opt) }
    modifiedAfter.foreach { opt => dfReader = dfReader.option("modifiedAfter", opt) }
    pathGlobFilter.foreach { opt => dfReader = dfReader.option("pathGlobFilter", opt) }

    dfReader.load(dir.getCanonicalPath)
  }
```

After adding these methods, this test could be changed to below:

```
  test(
    "SPARK-31962: when modifiedBefore specified" +
      " and sharing same timestamp with file last modified time.") {
    withTempDir { dir =>
      val time = LocalDateTime.now(ZoneOffset.UTC)
      val exc = intercept[AnalysisException] {
        executeTest(dir, time, modifiedBefore = Some(formatTime(time)))
      }
      assert(exc.getMessage.contains("Unable to infer schema for CSV"))
    }
  }
```